docs/devel/offline-virtual-machine.md

```diff
@@ -0,0 +1,234 @@
+
+Now its time to discuss the content of the OfflineVirtualMachine object.
+The object is defined in the same way as any other Kubernetes object. You can
+use YAML of JSON format to specify the content. The example structure is bellow:
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: OfflineVirtualMachine
+metadata:
+  name: myvm
+spec:
+  running: false
+  template:
+    metadata:
+      labels:
+        my: label
+    spec:
+      domain:
+        resources:
+          requests:
+            memory: 8Mi
+        devices:
+          disks:
+          - name: disk0
+            volumeName: mypcv
+      volumes:
+        - name: mypvc
+          persistentVolumeClaim:
+            claimName: myclaim
+```
+
+The file specification follows the Kubernetes guide. The apiVersion is linked
+with the Kubevirt release cycle.
+
+In the metadata section, there is a *required* field, the **name**. Then
+following the spec section, there are two important parts. The **running**, which
+indicates the current state of the VirtualMachine attached to this VirtualMachine.
+Second is the **template**, which is the VirtualMachine template.
+
+Let us go over each of these fields.
+
+### Name
+
+The name if important for the OfflineVirtualMachine, it is used to find created
+VirtualMachine. The VirtualMachine name is directly derived from
+the OfflineVirtualMachine. This means, if OfflineVirtualMachine is names 'testvm',
+the VirtualMachine is named 'testvm' too.
+
+When the VirtualMachineStatefulSet is created from the OfflineVirtualMachine,
+its name is also derived. In the case of 'testvm', name of each virtual machine
+in the stateful set is prefixed with the 'testvm' and appended with the ordering:
+'testvm0', 'testvm1' and so on.
+
+### Template
+
+The template is directly the VirtualMachine object. It contains the metadata
+section and a full VirtualMachineSpec.
+Thus it is identical to a VirtualMachine without a VirtualMachineStatus.
+This follows the same pattern which Kubernetes has on such template sections.
+There a PodTemplate contains of Metadata and a PodSpec.
+
+## OfflineVirtualMachine behaviour
+
+The OfflineVirtualMachine has to be in sync with its VirtualMachine. This means
+that the OfflineVirtualMachine has to observe both, the OfflineVirtualMachine
+and created VirtualMachine. When the link is established the config changes
+are translated to the VirtualMachine, and coresponding status changes are
+translated back to OfflineVirtualMachine.
+
+### Status
+
+The status of the VirtualMachine is watched and is reflected in OfflineVirtualMachine status.
+The status contains the current state of the VirtualMachine (running/not running)
+and whether it is ready.
+
+More detailed information is queried directly from the status of the VirtualMachine.
+
+### OwnerReference
+
+Linking the created VirtualMachine to its parent OfflineVirtualMachine pose
+a challenge. Using the same name is only part of the solution. To find the
```

Comment:
I find using names to uniquely identify something as very confusing and error prone. Do we want to guarantee its uniqueness and if so what would be the scope for it?
Shouldn't we use `uuid`s for it?

Reply:
You are right. I find it error prone too. There should be UUID. For the first implementation the names are sufficient, but for the next step we will better mechanism.

Reply:
Labels and names are the only ways in k8s to do something like that. Once you have fetched the object by name or label, you can check the OwnerReference to see if it is really adopted by the controller (but controller adopt their workloads also solely by namespace, name or labels and then place their OwnerReference on the object). We could make an exception just for OfflineVirtualMachine since it has a 1:1 relationship, but that is not how k8s works and it will not work for any other controller in k8s.

Reply:
You have the following guarantees:

 1) If there exists a VM with name X and it is not owned by an ovm, then there is no owner reference on it
 2) If a new ovm with name X is instantiated, it checks for vms with that name. If there exists a vm with name X and it has no owner reference, it will adopt it (add controller reference)
 3) If a new ovm with name X is instantiated and a vm with name X exists, and it is owned by another controller which still exists, it will not adopt it and wait until the situation changes. Until then it will periodically emit error events and reflect it in the status in a condition that it can't create a vm.

Reply:
I understand the limitations. Do we make sure that a name is unique and if so how do we namespace them?

Reply:
ovm and vm are in the same namespace. does that answer the question?

Reply:
so is it possible to use the same name of other ovm, vm pairs?

Reply:
> so is it possible to use the same name of other ovm, vm pairs?

in a different namespace yes. In the same namespace no. Either an ovm already exists, then I can't create one, or it does not. However a vm with a matching name could be created by anoter controller (not ovm). Therefore an owner reference of another controller type could be on the vm.

Reply:
In different namespace, it is possible. In the same namespace it will collide.

Reply:
I am looking for things that we need to validate. Thanks for clarification.